apache2/waf_logging/waf_log_util.cc

```diff
@@ -1,122 +1,269 @@
+        const char* details_file, const char* details_line, const char* hostname, const char* waf_unique_id) {
     
     try {
-        // Set Waf format.
-        json_string = get_json_log_message(resourceId, operationName, category, instanceId, clientIP, clientPort, requestUri, ruleSetType, ruleSetVersion, ruleId, messages, action, site, details_messages, details_data, details_file, details_line, hostname, waf_unique_id); 
-    }
-    catch (...) {
-        return WAF_LOG_UTIL_FAILED;
-    }
+        const std::string json_string = get_json_log_message(resource_id, operation_name, category,
+                instance_id, client_ip, client_port, request_uri, rule_set_type, rule_set_version,
+                rule_id, messages, action, site, details_messages, details_data, details_file,
+                details_line, hostname, waf_unique_id);
+        char* result = static_cast<char*>(std::malloc(json_string.length() + 2)); // Two extra bytes for \n and \0
+        std::copy(json_string.begin(), json_string.end(), result);
+        result[json_string.length()] = '\n';
+        result[json_string.length() + 1] = '\0';
+        return result;
+    } catch (...) {}
 
-    *result_json = strdup((json_string + "\n").c_str());
-    return WAF_LOG_UTIL_SUCCESS; 
+    return nullptr;
 }
 
 void free_json(char* str) {
-    free(str);
+    std::free(str);
 }
 
-void init_appgw_rules_id_hash() {
-	ifstream infile(RULES_ID_FILE);
-	string line;
-
-	while (getline(infile, line)) {
-		int rule_id = stoi(line);
-		appgw_ruleid_hash[rule_id] = true;
-	}
-	infile.close();
+namespace stdx = std::experimental;
+static stdx::optional<rule_ids_vector> read_disableable_rule_ids(const std::string& filename) {
+    std::ifstream ifs(filename);
+    if (ifs) {
+        rule_ids_vector rule_ids;
+        std::copy(std::istream_iterator<int>(ifs),
+                std::istream_iterator<int>(),
+                std::back_inserter(rule_ids));
+        std::sort(std::begin(rule_ids), std::end(rule_ids));
+        return rule_ids;
+    }
 
-	return;
+    return {};
 }
 
-bool rule_is_mandatory(int rule_id) {
-	return (appgw_ruleid_hash.find(rule_id) == appgw_ruleid_hash.end());
-}
\ No newline at end of file
+int init_appgw_rules_id_hash() {
+    const std::map<rule_set_info, std::string> disableableRulesPathsMap {
+        #define X(name) {OWASP_##name, "/etc/nginx/modsec/OWASP/""##name##""/disableableRules.conf"},
+        LIST_OF_VARIABLES
+        #undef X
+    };
+    for (const auto& item : disableableRulesPathsMap) {
+        const auto vec = read_disableable_rule_ids(item.second);
+        if (vec) {
+            disableable_rule_ids_map[item.first] = *vec;
+        }
+    }
+    if (disableable_rule_ids_map.empty()) {
+        // If we haven't found type/version-specific files, we need to read the default (deprecated one)
+        const auto vec = read_disableable_rule_ids("/etc/nginx/modsec/disableableRules.conf");
```

Comment:
could we put the path into header file?

Reply:
What would it buy us? It is not used anywhere outside of the `waf_log_util.cc`